### ISSUE TITLE
`sync_dist` works incorrectly when `self.log` gets key-value pairs in different order

### ISSUE THREAD

**Issue by iwyoo** [PyTorchLightning/pytorch-lightning#20946]
### Bug description

It seems that `self.log` with `sync_dist` option computes average score incorrectly if key-value pairs are digested in different order.

### What version are you seeing the problem on?

v2.3.3 (but v2.5 also gives the same result)

### How to reproduce the bug

To reproduce:
```python
import torch
import torch.nn as nn
import pytorch_lightning as pl
from pytorch_lightning import Trainer
from pytorch_lightning.strategies import DDPStrategy
from torch.utils.data import Dataset, DataLoader


class DummyDataset(Dataset):
    def __len__(self):
        return 8

    def __getitem__(self, idx):
        return torch.tensor([idx], dtype=torch.float)


class SimpleModel(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.dummy = nn.Linear(1, 1)

    def training_step(self, batch, batch_idx):
        rank = self.global_rank
        input_tensor = torch.tensor([[float(rank)]], device=self.device)
        output = self.dummy(input_tensor)

        self.log("rank_metric_sync", input_tensor, sync_dist=True)
        self.log("rank_metric_nosync", input_tensor, sync_dist=False)

        return output.sum()

    def test_step(self, batch, batch_idx):
        pass

    def on_test_epoch_end(self):
        # Working correctly
        self.log("test_rank_metric_sync", float(self.global_rank), sync_dist=True)
        self.log("test_rank_metric_nosync", float(self.global_rank), sync_dist=False)

        # Not working correctly
        if self.global_rank == 0:
            d = {}
            for i in range(10):
                d[str(i)] = float(i)
        else:
            d = {}
            for i in range(9, -1, -1):
                d[str(i)] = float(i)

        print(d)
        for key, value in d.items():
            self.log(f"test_dict_{key}", value, sync_dist=True)

    def configure_optimizers(self):
        return torch.optim.SGD(self.parameters(), lr=0.01)


if __name__ == "__main__":
    model = SimpleModel()
    dataset = DummyDataset()
    loader = DataLoader(dataset, batch_size=2)

    trainer = Trainer(
        accelerator="gpu",
        devices=2,
        max_epochs=1,
        strategy=DDPStrategy(find_unused_parameters=False),
        log_every_n_steps=1,
    )

    trainer.fit(model, loader)
    trainer.test(model, loader)
```

Two dictionaries:
```
{'9': 9.0, '8': 8.0, '7': 7.0, '6': 6.0, '5': 5.0, '4': 4.0, '3': 3.0, '2': 2.0, '1': 1.0, '0': 0.0}
{'0': 0.0, '1': 1.0, '2': 2.0, '3': 3.0, '4': 4.0, '5': 5.0, '6': 6.0, '7': 7.0, '8': 8.0, '9': 9.0}
```

Test metric:
```
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│        test_dict_0        │            4.5            │
│        test_dict_1        │            4.5            │
│        test_dict_2        │            4.5            │
│        test_dict_3        │            4.5            │
│        test_dict_4        │            4.5            │
│        test_dict_5        │            4.5            │
│        test_dict_6        │            4.5            │
│        test_dict_7        │            4.5            │
│        test_dict_8        │            4.5            │
│        test_dict_9        │            4.5            │
│  test_rank_metric_nosync  │            0.0            │
│   test_rank_metric_sync   │            0.5            │
└───────────────────────────┴───────────────────────────┘
```
Every `test_dict_n` should give `n` instead of 4.5.

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0): 2.3.3
#- PyTorch Version (e.g., 2.5): 2.0.1
#- Python version (e.g., 3.12): 3.10.11
#- OS (e.g., Linux): Ubuntu 22.04.3 LTS
#- CUDA/cuDNN version: CUDA 11.8
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

cc @justusschock

**Comment by rittik9** [comment]
HI @iwyoo thank you for raising this issue, will take a look

**Comment by SkafteNicki** [comment]
Just to chip in here, this happens because there are an implicit assumption that all ranks log to the same key in the same order. If we dig into the code, this is where the sync is started:
https://github.com/Lightning-AI/pytorch-lightning/blob/25b1343f1c112f6f8a87e07b336c1899c4065761/src/lightning/pytorch/trainer/connectors/logger_connector/result.py#L419-L423
if we call `print(self.rank, result_metric, value)` for the first case of difference order in the dicts we get something like
```
1 _ResultMetric('test_dict_2', value=0.0, cumulated_batch_size=0) tensor(2.)
0 _ResultMetric('test_dict_0', value=0.0, cumulated_batch_size=0) tensor(0.)
```
we can see that the two result metrics objects are correctly called something differently but after this point it does not matter, when `result_metric.forward(value, batch_size)` is called the synchronization implicitly assumes that the two objects are connected because they are called at the same time. Specifically for ddp the synchronization it is done by this function:
https://github.com/Lightning-AI/pytorch-lightning/blob/25b1343f1c112f6f8a87e07b336c1899c4065761/src/lightning/fabric/utilities/distributed.py#L183
which only sees the input tensor passed to the `forward` not the name. In the end this is really just how synchronization works, that the order of operations determines which values should be synced.

That said, I see 2 paths forward:
* Either be very clear that users should log to the same key on all devices in the same order
* If `sync_dist=True` also sync the `key` across devices and check that it is consistent before sync of value else throw an error

**Comment by SkafteNicki** [comment]
Going to close issue. The overall conclusion is that the base assumption of DDP, that code executed on two processes belong together leads to the implicit assumption that keys are the same across processes. Ping and reopen if necessary.
